src/solvers/mp/MasterProblem.cpp

```diff
@@ -0,0 +1,1278 @@
+		pNurse->roster_.reset();
+
+	for(MyVar* var: pModel_->getActiveColumns()){
+		if(pModel_->getVarValue(var) > EPSILON){
+		  PPattern pat = getPattern(var->getPattern());
+			LiveNurse* pNurse = theLiveNurses_[pat->nurseId_];
+			for(int k=pat->firstDay_; k<pat->firstDay_+pat->length_; ++k){
+        int s = pat->getShift(k);
+        if(s == 0) continue; // nothing to do
+        // assign a skill to the nurse for the shift
+				bool assigned = false;
+				for(int sk=0; sk<pScenario_->nbSkills_; ++sk)
+					if(skillsAllocation[k][s-1][sk][pNurse->pPosition_->id_] > EPSILON){
+						pNurse->roster_.assignTask(k,s,sk);
+						skillsAllocation[k][s-1][sk][pNurse->pPosition_->id_] --;
+						assigned = true;
+						break;
+					}
+				if(!assigned){
+					char error[255];
+					sprintf(error, "No skill found for Nurse %d on day %d on shift %d", pNurse->id_, k, s);
+					Tools::throwError((const char*) error);
+				}
+			}
+		}
+	}
+
+	//build the states of each nurse
+	solution_.clear();
+	for(LiveNurse* pNurse: theLiveNurses_){
+		pNurse->buildStates();
+		solution_.push_back(pNurse->roster_);
+	}
+}
+
+//------------------------------------------------------------------------------
+
+//------------------------------------------------------------------------------
+
+void MasterProblem::save(vector<int>& weekIndices, string outfile){
+	storeSolution();
+	// initialize the log stream
+	// first, concatenate the week numbers
+	int nbWeeks = weekIndices.size();
+
+	// write separately the solutions of each week in the required output format
+	int firstDay = pDemand_->firstDay_;
+	for(int w=0; w<nbWeeks; ++w){
+		//		string solutionFile = outdir+std::to_string(weekIndices[w])+".txt";
+		Tools::LogOutput solutionStream(outfile);
+		solutionStream << solutionToString(firstDay, 7, pScenario_->thisWeek()+w);
+		firstDay += 7;
+	}
+}
+
+//------------------------------------------------------------------------------
+// Build the, possibly fractional, roster corresponding to the solution
+// currently stored in the model
+//------------------------------------------------------------------------------
+vector3D<double> MasterProblem::getFractionalRoster() {
+	vector3D<double> fractionalRoster;
+  Tools::initVector3D(fractionalRoster, getNbNurses(), getNbDays(), pDemand_->nbShifts_, .0);
+
+	// Retrieve current fractional roster for each nurse
+	for(MyVar* var : pModel_->getActiveColumns()){
+		if (var->getPattern().empty()) continue;
+    double value = pModel_->getVarValue(var);
+		if(value < EPSILON) continue;
+		PPattern pat = getPattern(var->getPattern());
+    vector2D<double>& fractionalRoster2 = fractionalRoster[pat->nurseId_];
+		for(int k=pat->firstDay_; k<pat->firstDay_+pat->length_; ++k)
+			fractionalRoster2[k][pat->getShift(k)] += value;
+	}
+
+	return fractionalRoster;
+}
+
+void MasterProblem::checkIfPatternAlreadyPresent(const std::vector<double>& pattern) const {
+  for(MyVar* var: pModel_->getActiveColumns()) {
```

Comment:
I may be confused, but it seems the function returns as soon as it find a pattern different from the input one. It should go through all active columns before returning, shouldnt it?

Reply:
True, fix with 604c250dd7ea93b1199521c191769becae7e37ae